src/Acl.php

```diff
@@ -45,9 +49,40 @@ public function getRules(Model $model)
                 $modelClasses[] = $class;
             }
         } while (($class = get_parent_class($class)) !== false);
-        $res = $user->ref('AccessRules')->addCondition('model', 'in', $modelClasses);
 
-        return $res; // @phpstan-ignore-line
+        $this->skipApplyRestrictionsForRulesExport = true;
+        try {
+            $rules = $user->ref('AccessRules')
+                ->addCondition('model', 'in', $modelClasses)
+                ->export(['model', 'all_visible', 'visible_fields', 'all_editable', 'editable_fields', 'all_actions', 'actions', 'conditions']);
+
+            // normalize
+            foreach ($rules as $k => $rule) {
+                $rules[$k]['visible_fields'] = $rule['all_visible'] ? [] : $this->normalizeValue($rule['visible_fields']);
+                $rules[$k]['editable_fields'] = $rule['all_editable'] ? [] : $this->normalizeValue($rule['editable_fields']);
+                $rules[$k]['actions'] = $rule['all_actions'] ? [] : $this->normalizeValue($rule['actions']);
+            }
+        } finally {
+            $this->skipApplyRestrictionsForRulesExport = false;
+        }
+
+        return $rules;
+    }
+
+    /**
+     * @param list<string>|string|null $v
+     *
+     * @return list<string>
+     */
+    private function normalizeValue($v): array
+    {
+        if (($v ?? '') === '') {
+            return [];
+        }
+
+        return is_array($v)
+            ? $v
+            : explode(',', $v);
```

Comment:
```suggestion
    /**
     * @return list<string>
     */
    private function normalizeValue(?string $v): array
    {
        return ($v ?? '') === ''
            ? []
            : explode(',', $v);
```

based on https://github.com/atk4/login/blob/edec0f86a1868f40b12d3f37bcf9aed0610a73be/src/Model/AccessRule.php#L55 it cannot be an array - implicit Field type is always a `string`